README.rst
PyCon UK Scheduler
==================
A tool to produce the schedule for the
`PyCon UK Conference <http://pyconuk.org>`_.

Installation
============

The library requires Python 3.6 or later. The simplest way to install is::

    pip install pyconuk-scheduler

To install from source code::

    git clone https://github.com/PyConUK/PyConUKScheduler
    cd PyConUKScheduler
    python setup.py install

Quick Start
===========

The tool expects to find a directory with the necessary data files to define
the conference. By default it will look for a directory named `input` under
the current working directory.

It will generate YAML files in the format required by the conference
website. By default, these will be placed a in a `build` directory under the
current working directory.

It will also generate `.csv` and `.pickle` files to store the calculated
schedule and its associated definition and a log file with the full details of
the most recent calculation. By default, these will be placed in a
`solution` directory under the current working directory.

The `input`, `solution` and `build` directories can be passed to the tool as
command line options in place of the default locations.

To build the schedule with the default options::

    scheduler build

There are also further options, e.g. to set the logging verbosity or the
solver algorithm to use. To view those options::

    scheduler --help

or, for the options for a specifc command::

    scheduler build --help


CHANGELOG.rst
Changelog
#########

v0.10.1 (2017-10-02)
--------------------
* Minor: Fix line ending of schedule.csv

v0.10.0 (2017-09-16)
--------------------
* Breaking Change: Output markdown files are no longer produced

* New Feature: The ability to add 'unbounded' events which occur in all rooms
  on all days

* Minor: Speaker names in their original form are included within the pickled
  events dictionary

* Minor: The full slots list, with index number, is included in the log file

* Minor: Allocated and unbounded events now have tags added automatically
  which ensure uniqueness (no longer necessary to use e.g. 'morning' tag)

* Bug Fix: Durations for unbounded and allocated events are taken from their
  definition (previously hard coded as zero)

v0.9.1 (2017-09-06)
-------------------
* Minor: io functions now take the working directories as arguments

* Minor: All proposals now included in pickled resources

v0.9.0 (2017-09-06)
-------------------
* New Feature: --reload option on validate command to control whether the
  conference definition is reloaded from input YAML files

* Bug Fix: --help option threw an error if run outside a valid project folder
  structure

v0.8.4 (2017-09-04)
-------------------
* Minor: Subtitles from proposals are now imported

v0.8.3 (2017-09-02)
-------------------
* Bug Fix: Original schedule and new schedule were being passed to the diff
  function in the wrong order

v0.8.2 (2017-09-01)
-------------------
* Bug Fix: Pre-allocated events were being scheduled in addition to their
  pre-allocated slot when optimising for consistency

v0.8.1 (2017-08-24)
-------------------
* Minor: Order of fields altered in schedule.csv

v0.8.0 (2017-08-21)
-------------------
* New Feature: --diff option to control logging of schedule diff


v0.7.2 (2017-08-21)
-------------------
* Bug Fix: Consistency optimisation fixed

v0.7.1 (2017-08-21)
-------------------
* Minor: Updated log message for Unavailability

v0.7.0 (2017-08-21)
-------------------
* Bug Fix: Remove use of strings for Slot.starts_at

* New Feature: Unavailability and clashes can now be defined for events as well
  as people

* New feature: Capacity objective now split into efficiency or equity

v0.6.0 (2017-08-18)
-------------------
* New Feature: Pre-allocated events can be defined in allocations.yaml

* Breaking Change: venues.yml renamed to timetable.yml

v0.5.0 (2017-08-17)
-------------------
* Breaking Change: content previously in separate yaml files now in venues.yaml

* New Feature: objective function argument added to build command

v0.4.1 (2017-08-16)
-------------------
* Improved logging with counts of events and slots by event type

v0.4.0 (2017-08-14)
-------------------
* New Feature: Venue availability can be defined

* Minor Improvement: Order of columns in schedule.csv swapped to show index
  numbers at lhs of file

v0.3.1 (2017-08-10)
-------------------
* Bug Fix: Redundant export of schedule on rebuild removed

v0.3.0 (2017-08-10)
-------------------
* New feature: Rebuild command
  Rebuild the output files from a previously computed schedule

v0.2.0 (2017-08-10)
-------------------
* New Featue: Validate command
  Validate a previously computed schedule

v0.1.0 (2017-08-10)
-------------------
* Initial release


tests/test_denormalise.py
from datetime import datetime

import pytest
from conference_scheduler.resources import Event
from conference_scheduler.resources import Slot

import scheduler.denormalise as dn


@pytest.fixture
def event_types():
    return ['talk', 'workshop', 'plenary']


@pytest.fixture
def venues():
    return {
        'Assembly Room': {
            datetime(2017, 10, 26, 0, 0): {
                'morning': [
                    {
                        'starts_at': 36900,
                        'duration': 30,
                        'event_type': 'plenary',
                        'capacity': 600
                    },
                    {
                        'starts_at': 39600,
                        'duration': 30,
                        'event_type': 'plenary',
                        'capacity': 600
                    },
                ],
                'afternoon': [
                    {
                        'starts_at': 52200,
                        'duration': 30,
                        'event_type': 'talk',
                        'capacity': 600
                    },
                    {
                        'starts_at': 54000,
                        'duration': 30,
                        'event_type': 'talk',
                        'capacity': 600
                    }
                ]
            }
        }
    }


@pytest.fixture
def events_definition():
    return [
        {
            'title': 'UKPA AGM',
            'duration': 45,
            'tags': ['plenary'],
            'person': 'owen-campbell',
            'event_type': 'talk',
            'demand': 0
        },
        {
            'title': 'A very exciting talk',
            'duration': 30,
            'tags': ['talk'],
            'person': 'owen-campbell',
            'event_type': 'talk',
            'demand': 0
        },
        {
            'title': 'A very interesting talk',
            'duration': 30,
            'tags': ['talk'],
            'person': 'vincent-knight',
            'event_type': 'talk',
            'demand': 0
        },
        {
            'title': 'A slightly dull talk',
            'duration': 30,
            'tags': [],
            'person': 'joe-bloggs',
            'event_type': 'talk',
            'demand': 0
        },
        {
            'title': 'A fascinating workshop',
            'duration': 90,
            'tags': ['talk'],
            'person': 'alice',
            'event_type': 'workshop',
            'demand': 0
        },
        {
            'title': 'A beginner workshop',
            'duration': 90,
            'tags': ['talk'],
            'person': 'bob',
            'event_type': 'workshop',
            'demand': 0
        }
    ]


@pytest.fixture
def people_unavailability_defintion():
    return {
        'owen-campbell': [{
            'unavailable_from': datetime(2017, 10, 26, 0, 0),
            'unavailable_until': datetime(2017, 10, 26, 23, 59)
        }]
    }


@pytest.fixture
def events_unavailability_defintion():
    return {
        'A very interesting talk': [{
            'unavailable_from': datetime(2017, 10, 26, 0, 0),
            'unavailable_until': datetime(2017, 10, 26, 11, 59)
        }]
    }


@pytest.fixture
def people_clashes_definition():
    return {
        'owen-campbell': ['vincent-knight', 'joe-bloggs'],
        'vincent-knight': ['owen-campbell']
    }


@pytest.fixture
def events_clashes_definition():
    return {
        'A slightly dull talk': ['A very interesting talk'],
    }


@pytest.fixture
def allocations_definition():
    return [
        {'Another talk': {
            'person': 'charlie',
            'tags': None,
            'venue': 'Assembly Room',
            'day': datetime(2017, 10, 26, 0, 0),
            'session': 'afternoon',
            'starts_at': 54000}},
        {'Another workshop': {
            'person': 'diane',
            'tags': None,
            'venue': 'Room C',
            'day': datetime(2017, 10, 26, 0, 0),
            'session': 'afternoon',
            'starts_at': 54000}}]


def test_types_and_slots(event_types, venues):
    types_and_slots = dn.types_and_slots(venues)
    assert all(isinstance(item['slot'], Slot) for item in types_and_slots)
    assert len(types_and_slots) == 4


def test_types_and_events(events_definition):
    types_and_events = dn.types_and_events(events_definition)
    assert all(isinstance(item['event'], Event)for item in types_and_events)
    assert len(types_and_events) == 6


def test_people_unavailability(
    event_types, events_definition, venues, people_unavailability_defintion
):
    types_and_slots = dn.types_and_slots(venues)
    slots = [item['slot'] for item in types_and_slots]
    unavailability = dn.people_unavailability(
        events_definition, slots, people_unavailability_defintion)
    print(unavailability)
    assert unavailability == {0: [0, 1, 2, 3], 1: [0, 1, 2, 3]}


def test_events_unavailability(
    event_types, events_definition, venues, events_unavailability_defintion
):
    types_and_slots = dn.types_and_slots(venues)
    slots = [item['slot'] for item in types_and_slots]
    unavailability = dn.events_unavailability(
        events_definition, slots, events_unavailability_defintion)
    print(unavailability)
    assert unavailability == {2: [0, 1]}


def test_people_clashes(events_definition, people_clashes_definition):
    clashes = dn.people_clashes(events_definition, people_clashes_definition)
    assert clashes == ({0: [2, 3, 1], 1: [2, 3, 0], 2: [0, 1], 3: [], 4: [], 5: []}, 2)


def test_events_clashes(events_definition, events_clashes_definition):
    clashes = dn.events_clashes(events_definition, events_clashes_definition)
    assert clashes == ({3: [2]})


def test_unsuitability(venues, events_definition):
    types_and_slots = dn.types_and_slots(venues)
    unsuitability = dn.unsuitability(types_and_slots, events_definition)
    assert list(unsuitability.keys()) == list(range(6))


def test_allocations(allocations_definition):
    allocations = dn.allocations(allocations_definition)
    assert len(allocations) == len(allocations_definition)
    assert all(isinstance(item['event'], Event) for item in allocations)
    assert all(isinstance(item['slot'], Slot) for item in allocations)


src/scheduler/io.py
import csv
import pickle
from pathlib import Path
from pprint import pformat

import daiquiri
from conference_scheduler import converter
from ruamel.yaml import YAML
from slugify import slugify

logger = daiquiri.getLogger(__name__)
yaml = YAML(typ='safe')
yaml.default_flow_style = False


def import_yaml(input_folder):
    """Import all yaml files in the given folder into a single resources
    dict"""
    yaml_resources = {}
    yaml_files = [
        path for path in input_folder.iterdir()
        if path.suffix == '.yml']
    for path in yaml_files:
        with path.open('r') as file:
            yaml_resources[path.stem] = yaml.load(file)
    logger.debug(f'\nreources:\n{pformat(yaml_resources)}')
    return yaml_resources


def import_proposals(resources, input_folder):
    """Import the proposals data from a .csv file"""
    proposals = []
    with Path(input_folder, 'proposals.csv').open('r') as file:
        reader = csv.DictReader(file)
        for row in reader:
            event_type = row['session_type']
            proposals.append({
                'title': row['title'],
                'duration': int(row['duration']),
                'demand': float(row.get('demand', 0)),
                'person': slugify(row['name']),
                'name': row['name'],
                'tags': [row['tag']] if row['tag'] != '' else [],
                'subtitle': row['subtitle'],
                'description': row['description'],
                'event_type': event_type})
    logger.debug(f'\nreources:\n{pformat(proposals)}')
    return proposals


def import_solution(solution_folder):
    """Import a previously computed schedule from a .csv file"""
    csv_file = Path(solution_folder, 'schedule.csv')
    logger.info(f'Importing schedule from {csv_file}')
    solution = []
    with Path(csv_file).open('r') as file:
        reader = csv.DictReader(file)
        for row in reader:
            solution.append((
                int(row['event_index']),
                int(row['slot_index'])))
        logger.debug(f'\nreources:\n{pformat(solution)}')
    return solution


def import_schedule_definition(solution_folder):
    """Import previously pickled schedule"""
    pickle_file = Path(solution_folder, 'scheduler.pickle')
    logger.info(
        f'Importing resources, events, slots and schedule from {pickle_file}')
    with pickle_file.open('rb') as f:
        bundle = pickle.load(f)
    return bundle


def pickle_solution_and_definition(
    resources, events, slots, allocations, solution, solution_folder
):
    """Store the computed solution, the resources dict and the associated
    events and slots lists in pickle format"""
    pickle_file = Path(solution_folder, 'scheduler.pickle')
    logger.info(
        f'Pickling resources, events, slots and schedule to {pickle_file}')
    bundle = {
        'resources': resources,
        'events': events,
        'slots': slots,
        'allocations': allocations,
        'solution': solution
    }
    with pickle_file.open('wb') as f:
        pickle.dump(bundle, f, pickle.HIGHEST_PROTOCOL)


def export_schedule(solution, events, slots, solution_folder):
    """Write a human readable .csv file of the computed solution"""
    csv_file = Path(solution_folder, 'schedule.csv')
    logger.info(f'Exporting schedule to {csv_file}')

    schedule = converter.solution_to_schedule(solution, events, slots)
    scheduled_items = [
        {
            'event_index': events.index(item.event),
            'event': f'{item.event.name}',
            'slot_index': slots.index(item.slot),
            'slot': f'{item.slot.starts_at} {item.slot.venue}'
        }
        for item in schedule
    ]
    with csv_file.open('w') as f:
        writer = csv.DictWriter(
            f, fieldnames=['event_index', 'event', 'slot_index', 'slot'])
        writer.writeheader()
        for item in scheduled_items:
            writer.writerow(item)


def export_solution_and_definition(
    resources, events, slots, allocations, solution, solution_folder
):
    solution_folder.mkdir(exist_ok=True)
    pickle_solution_and_definition(
        resources, events, slots, allocations, solution, solution_folder)
    export_schedule(solution, events, slots, solution_folder)


src/scheduler/cli.py
"""Procedures to define the Command Line Interface (cli)"""
from pathlib import Path

import click
from conference_scheduler.scheduler import event_schedule_difference
from conference_scheduler.converter import solution_to_schedule
from conference_scheduler.validator import (
    is_valid_solution, solution_violations)
import daiquiri

import scheduler.calculate as calc
from scheduler.decorators import timed
import scheduler.define as defn
from scheduler import convert, io, logging, session

logger = daiquiri.getLogger(__name__)


def events_and_slots(resources):
    slots = defn.slots(resources)
    events = defn.events(resources)
    unavailability = defn.unavailability(resources, slots)
    clashes = defn.clashes(resources)
    unsuitability = defn.unsuitability(resources, slots)

    defn.add_unavailability_to_events(events, slots, unavailability)
    defn.add_clashes_to_events(events, clashes)
    defn.add_unsuitability_to_events(events, slots, unsuitability)
    return events, slots


@click.version_option(message='%(prog)s %(version)s :: UK Python Association')
@click.group()
@click.option(
    '--verbosity', '-v', default='info',
    type=click.Choice(['critical', 'error', 'warning', 'info', 'debug']),
    help='Logging verbosity')
def scheduler(verbosity):
    pass


@scheduler.command()
@click.option(
    '--verbosity', '-v', default='info',
    type=click.Choice(['critical', 'error', 'warning', 'info', 'debug']),
    help='Logging verbosity')
@click.option(
    '--algorithm', '-a', default='pulp_cbc_cmd',
    type=click.Choice(
        ['pulp_cbc_cmd', 'glpk', 'hill_climber', 'simulated_annealing']),
    help='Solver algorithm')
@click.option(
    '--objective', '-o', default=None,
    type=click.Choice(['efficiency', 'equity', 'consistency']),
    help='Objective Function')
@click.option('--diff/--no-diff', default=False, help='Show schedule diff')
@click.option(
    '--input_dir', '-i', default=None, help='Directory for input files')
@click.option(
    '--solution_dir', '-s', default=None, help='Directory for solution files')
@click.option(
    '--build_dir', '-b', default=None, help='Directory for output yaml files')
@timed
def build(
    verbosity, algorithm, objective, diff, input_dir, solution_dir, build_dir
):
    logging.setup(verbosity)
    if input_dir:
        session.folders['input'] = Path(input_dir)

    if solution_dir:
        session.folders['solution'] = Path(solution_dir)

    if build_dir:
        session.folders['build'] = Path(build_dir)

    resources = defn.resources()
    events, slots = events_and_slots(resources)

    slots_by_index = {
        idx: f'{slot.starts_at} {slot.venue}'
        for idx, slot in enumerate(slots)}
    logger.debug(f'\nSlots List:\n{slots_by_index}')

    kwargs = {}
    if objective == 'consistency' or diff:
        original_solution = io.import_solution(session.folders['solution'])
        revised_solution = [
            item for item in original_solution
            if item[0] < len(events)]
        original_schedule = solution_to_schedule(
            revised_solution, events, slots)

    if objective == 'consistency':
        diff = True
        kwargs['original_schedule'] = original_schedule

    solution = calc.solution(events, slots, algorithm, objective, **kwargs)

    if diff:
        schedule = solution_to_schedule(solution, events, slots)
        event_diff = event_schedule_difference(original_schedule, schedule)
        logger.debug(f'\nevent_diff:')
        for item in event_diff:
            logger.debug(f'{item.event.name} has moved from {item.old_slot.venue} at {item.old_slot.starts_at} to {item.new_slot.venue} at {item.new_slot.starts_at}')

    if solution is not None:
        allocations = defn.allocations(resources)
        unbounded = defn.unbounded(resources)
        defn.add_allocations(events, slots, solution, allocations, unbounded)
        logger.debug(convert.schedule_to_text(solution, events, slots))
        io.export_solution_and_definition(
            resources, events, slots, allocations, solution,
            session.folders['solution'])


@scheduler.command()
@click.option(
    '--verbosity', '-v', default='info',
    type=click.Choice(['critical', 'error', 'warning', 'info', 'debug']),
    help='Logging verbosity')
@click.option(
    '--input_dir', '-i', default=None, help='Directory for input files')
@click.option(
    '--solution_dir', '-s', default=None, help='Directory for solution files')
@click.option(
    '--reload/--no-reload', default=False, help='Reload YAML definition')
@timed
def validate(verbosity, input_dir, solution_dir, reload):
    logging.setup(verbosity)
    if solution_dir:
        session.folders['solution'] = Path(solution_dir)

    solution = io.import_solution(session.folders['solution'])

    if reload:
        resources = defn.resources()
        events, slots = events_and_slots(resources)
        original_solution = io.import_solution(session.folders['solution'])
        solution = [
            item for item in original_solution
            if item[0] < len(events)]
    else:
        solution = io.import_solution(session.folders['solution'])
        definition = io.import_schedule_definition(session.folders['solution'])
        events = definition['events']
        slots = definition['slots']

    logger.info('Validating schedule...')
    if is_valid_solution(solution, events, slots):
        logger.info('Imported solution is valid')
    else:
        for v in solution_violations(
                solution, definition['events'], definition['slots']):
            logger.error(v)


src/scheduler/calculate.py
import daiquiri
from conference_scheduler import scheduler
from conference_scheduler.heuristics import hill_climber
from conference_scheduler.heuristics import simulated_annealing
from conference_scheduler.lp_problem.objective_functions import (
    capacity_demand_difference, number_of_changes)
from pulp import GLPK
from pulp import PULP_CBC_CMD

logger = daiquiri.getLogger(__name__)

solvers = {
    'pulp_cbc_cmd': {
        'function': scheduler.solution,
        'kwargs': {
            'solver': PULP_CBC_CMD(msg=False)}},
    'glpk': {
        'function': scheduler.solution,
        'kwargs': {
            'solver': GLPK(msg=False)}},
    'hill_climber': {
        'function': scheduler.heuristic,
        'kwargs': {
            'algorithm': hill_climber}},
    'simulated_annealing': {
        'function': scheduler.heuristic,
        'kwargs': {
            'algorithm': simulated_annealing}}}

objectives = {
    'capacity': capacity_demand_difference,
    'consistency': number_of_changes}


def solution(events, slots, solver, objective=None, **kwargs):
    message = f'Scheduling conference using {solver} solver'
    if objective is not None:
        message += f' and optimising {objective}'
    logger.info(f'{message}...')

    objective_function = objectives.get(objective, None)
    full_kwargs = {
        'events': events,
        'slots': slots,
        'objective_function': objective_function,
        **solvers[solver]['kwargs'],
        **kwargs}

    try:
        solution = solvers[solver]['function'](**full_kwargs)
    except ValueError:
        logger.error('No valid solution found')
        solution = None

    return solution


src/scheduler/denormalise.py
"""Functions which translate the human readable, nested definition data from
the input yaml files into the flattened structures required by the conference
scheduler computation engine."""

from datetime import datetime
from datetime import timedelta

from conference_scheduler.resources import Event
from conference_scheduler.resources import Slot


def types_and_slots(timetable):
    """
    Parameters
    ----------
    timetable : dict
        of the form
            {
                <venue name>: {
                    <datetime of a day>: {
                        <session name>: {
                            'starts_at': <seconds after midnight>,
                            'duration': <minutes>,
                            'event_type': <event type>,
                            'capacity': <capacity
                        }
                    }
                }
            }
    Returns
    -------
    list
        of dicts mapping an event type to a Slot instance
    """
    return [
        {
            'event_type': slot['event_type'],
            'slot': Slot(
                venue=venue,
                starts_at=(
                    datetime.combine(day, datetime.min.time()) +
                    timedelta(seconds=slot['starts_at'])),
                duration=slot.get('duration', 0),
                session=f'{day} {session}',
                capacity=slot.get('capacity', 0))
        }
        for venue, days in timetable.items()
        for day, sessions in days.items()
        for session, slots in sessions.items()
        for slot in slots
    ]


def types_and_events(events_definition):
    """
    Parameters
    ----------
    events_definition : list
        of dicts of the form
            {'title': Event title,
            'duration': <integer in minutes>,
            'tags': <list of strings>,
            'person': <string>}
    Returns
    -------
    list
        of dicts mapping an event type to an Event instance
    """
    return [
        {
            'event_type': event['event_type'],
            'event': Event(
                name=event['title'],
                duration=event['duration'],
                demand=event['demand'],
                tags=event['tags'])
        }
        for event in events_definition]


def people_unavailability(events_definition, slots, unavailability_definition):
    """
    Parameters
    ----------
    events_definition : list
        of dicts of the form
            {'title': Event title,
            'duration': <integer in minutes>,
            'tags': <list of strings>,
            'person': <string>,
            'event_type': <string>}
    slots : list
        of Slot instances
    unavailablity_definition : dict
        mapping a person to a list of time periods. e.g.
            {'owen-campbell': [{
                'unavailable_from': datetime(2017, 10, 26, 0, 0),
                'unavailable_until': datetime(2017, 10, 26, 23, 59)}]
            }

    Returns
    -------
    dict
        mapping the index of an event in the events list to a list of slots
        for which it must not scheduled. The slots are represented by their
        index in the slots list.
    """
    return {
        events_definition.index(event): [
            slots.index(slot)
            for period in periods
            for slot in slots
            if period['unavailable_from'] <= slot.starts_at and
            period['unavailable_until'] >= (
                slot.starts_at + timedelta(0, slot.duration * 60))
        ]
        for person, periods in unavailability_definition.items()
        for event in events_definition if event['person'] == person
    }


def events_unavailability(events_definition, slots, unavailability_definition):
    """
    Parameters
    ----------
    events_definition : list
        of dicts of the form
            {'title': Event title,
            'duration': <integer in minutes>,
            'tags': <list of strings>,
            'person': <string>,
            'event_type': <string>}
    slots : list
        of Slot instances
    unavailablity_definition : dict
        mapping an event to a list of time periods. e.g.
            {'An Interesting Talk': [{
                'unavailable_from': datetime(2017, 10, 26, 0, 0),
                'unavailable_until': datetime(2017, 10, 26, 23, 59)}]
            }

    Returns
    -------
    dict
        mapping the index of an event in the events list to a list of slots
        for which it must not scheduled. The slots are represented by their
        index in the slots list.
    """
    return {
        events_definition.index(event): [
            slots.index(slot)
            for period in periods
            for slot in slots
            if period['unavailable_from'] <= slot.starts_at and
            period['unavailable_until'] >= (
                slot.starts_at + timedelta(0, slot.duration * 60))
        ]
        for event_title, periods in unavailability_definition.items()
        for event in events_definition if event['title'] == event_title
    }


def people_clashes(events_definition, clashes_definition):
    """
     Parameters
    ----------
    events_definition : list
        of dicts of the form
            {'title': Event title,
            'duration': <integer in minutes>,
            'tags': <list of strings>,
            'person': <string>,
            'event_type': <string>}
    clashes_definition : dict
        mapping a person to a list of people whose events they must not not be
        scheduled against.

    Returns
    -------
    dict
        mapping the index of an event in the events list to a list of event
        indexes against which it must not be scheduled.
    integer
        the count of self-clashes added
    """
    # Add everyone who is missing to the clashes definition so that they cannot
    # clash with themselves
    for person in [event['person'] for event in events_definition]:
        if person not in clashes_definition:
            clashes_definition[person] = [person]

    # Add the self-clashing constraint to any existing entries where it is
    # missing
    count = 0
    for person, clashing_people in clashes_definition.items():
        if person not in clashing_people:
            clashing_people.append(person)
            count += 1

    clashes = {
        events_definition.index(event): [
            events_definition.index(t) for c in clashing_people
            for t in events_definition
            if t['person'] == c and
            events_definition.index(event) != events_definition.index(t)]
        for person, clashing_people in clashes_definition.items()
        for event in events_definition if event['person'] == person
    }
    return clashes, count


def events_clashes(events_definition, clashes_definition):
    """
     Parameters
    ----------
    events_definition : list
        of dicts of the form
            {'title': Event title,
            'duration': <integer in minutes>,
            'tags': <list of strings>,
            'person': <string>,
            'event_type': <string>}
    clashes_definition : dict
        mapping an event to a list of events they must not not be
        scheduled against.

    Returns
    -------
    dict
        mapping the index of an event in the events list to a list of event
        indexes against which it must not be scheduled.
    """
    clashes = {
        events_definition.index(event): [
            events_definition.index(t) for c in clashing_events
            for t in events_definition
            if t['title'] == c and
            events_definition.index(event) != events_definition.index(t)]
        for event_title, clashing_events in clashes_definition.items()
        for event in events_definition if event['title'] == event_title
    }
    return clashes


def unsuitability(types_and_slots, events_definition):
    """
    Parameters
    ----------
    types_and_slots : list
        of dicts mapping an event type to a Slot instance
    events_definition : list
        of dicts of the form
            {'title': Event title,
            'duration': <integer in minutes>,
            'tags': <list of strings>,
            'person': <string>,
            'event_type': <string>}
    Returns
    -------
    dict
        mapping the index of an event in the events list to a list of slots
        for which it must not scheduled. The slots are represented by their
        index in the slots list.
    """
    output = {}
    for i, event in enumerate(events_definition):
        unsuitable_slots = [
            i for i, dictionary in enumerate(types_and_slots)
            if dictionary['event_type'] != event['event_type']]
        output[i] = unsuitable_slots
    return output


def allocations(allocations_definition):
    try:
        return [
            {
                'event': Event(
                    name=event,
                    duration=details['duration'],
                    demand=0,
                    tags=details['tags']),
                'slot': Slot(
                    venue=details['venue'],
                    starts_at=(datetime.combine(
                        details['day'],
                        datetime.min.time()) +
                        timedelta(seconds=details['starts_at'])),
                    duration=details['duration'],
                    session=(f'{details["day"]} {details["session"]}'),
                    capacity=-1)
            }
            for allocation in allocations_definition
            for event, details in allocation.items()]
    except TypeError:
        return []


def unbounded(unbound_definition, timetable):
    return [
        {
            'event': Event(
                name=event,
                duration=details['duration'],
                demand=0,
                tags=[day, venue]),
            'slot': Slot(
                venue=venue,
                starts_at=(datetime.combine(
                    day,
                    datetime.min.time()) +
                    timedelta(seconds=details['starts_at'])),
                duration=details['duration'],
                session=f'{day} unbound',
                capacity=0)
        }
        for unbound in unbound_definition
        for event, details in unbound.items()
        for venue, days in timetable.items()
        for day in days
    ]
